File: src-tauri/src/app_config.rs
//! アプリ設定の単一ソース（[`AppConfig`]）と、ディスク上の `app.json` への読み書き。
//!
//! - **キー**: JSON のプロパティ名は各 struct のフィールドに対応させ、`#[serde(rename_all = "camelCase")]` で
//!   フロント（JS）とファイル形式を揃える。新しい設定はネストした struct を足す形で追加する。
//! - **読み書き**: 他コードからは [`load`] / [`save`] / [`mutate`] のみを使う（パスやファイル名はこのモジュール内だけ）。

use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

const CONFIG_FILENAME: &str = "app.json";

/// ディスクと IPC でやり取りする設定のルート。フィールドを増やすときはここ（と下位の struct）だけ触る。
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(default, rename_all = "camelCase")]
pub struct AppConfig {
    pub ui: UiSettings,
    pub connections: Vec<ConnectionProfile>,
}

impl Default for AppConfig {
    fn default() -> Self {
        Self {
            ui: UiSettings::default(),
            connections: vec![],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(default, rename_all = "camelCase")]
pub struct UiSettings {
    /// Whether the sidebar is expanded.
    pub sidebar_open: bool,
}

impl Default for UiSettings {
    fn default() -> Self {
        Self { sidebar_open: true }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConnectionProfile {
    pub id: String,
    pub label: String,
    pub host: String,
    pub port: u16,
    pub database: String,
    pub user: String,
    pub password: String,
    /// When `false`, password is never stored in the profile (empty string); prompt at connect time (future).
    #[serde(default = "default_save_password_in_profile")]
    pub save_password_in_profile: bool,
}

fn default_save_password_in_profile() -> bool {
    false
}

fn config_path(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = app.path().app_local_data_dir().map_err(|e| e.to_string())?;
    Ok(dir.join(CONFIG_FILENAME))
}

/// 設定ファイルを読む。無い場合は [`AppConfig::default`]。
pub fn load(app: &AppHandle) -> Result<AppConfig, String> {
    let path = config_path(app)?;
    if !path.exists() {
        return Ok(AppConfig::default());
    }
    let raw = fs::read_to_string(&path).map_err(|e| e.to_string())?;
    serde_json::from_str(&raw).map_err(|e| e.to_string())
}

/// 設定を丸ごと保存する。
pub fn save(app: &AppHandle, config: &AppConfig) -> Result<(), String> {
    let path = config_path(app)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let raw = serde_json::to_string_pretty(config).map_err(|e| e.to_string())?;
    fs::write(&path, raw).map_err(|e| e.to_string())
}

/// 読み込み → クロージャで変更 → 保存。Rust 側（将来のコマンドやプラグイン）から一部更新するときに使う。
// 現状フロントは `invoke(load/save)` のみ。未使用警告を抑止。
#[allow(dead_code)]
pub fn mutate<F>(app: &AppHandle, f: F) -> Result<AppConfig, String>
where
    F: FnOnce(&mut AppConfig),
{
    let mut config = load(app)?;
    f(&mut config);
    save(app, &config)?;
    Ok(config)
}

File: src/appConfig.js
/**
 * アプリ設定の読み書きはこのモジュール経由に統一する。
 * - IPC コマンド名は COMMANDS のみ
 * - オブジェクトの形は Rust の AppConfig と揃え、プロパティは camelCase（serde と一致）
 */

/** @typedef {{ sidebarOpen: boolean, sidebarWidthPx: number }} UiSettings */
/**
 * @typedef {{
 *   kind: 'default';
 * } | {
 *   kind: 'all';
 * } | {
 *   kind: 'selected';
 *   schemas: string[];
 * }} UserSchemaVisibility
 */
/**
 * @typedef {{
 *   id: string;
 *   label: string;
 *   host: string;
 *   port: number;
 *   database: string;
 *   user: string;
 *   password: string;
 *   savePasswordInProfile: boolean;
 *   userSchemaVisibility?: UserSchemaVisibility;
 * }} ConnectionProfile
 */
/** @typedef {{ ui: UiSettings, connections: ConnectionProfile[] }} AppConfig */

export const COMMANDS = {
  load: "load_app_config",
  save: "save_app_config",
};

function getInvoke() {
  return globalThis.__TAURI__?.core?.invoke;
}

export function defaultAppConfig() {
  return {
    ui: { sidebarOpen: true, sidebarWidthPx: 256 },
    connections: [],
  };
}

/** @returns {Promise<AppConfig | null>} */
export async function loadAppConfig() {
  const invoke = getInvoke();
  if (!invoke) return null;
  return invoke(COMMANDS.load);
}

export async function saveAppConfig(config) {
  const invoke = getInvoke();
  if (!invoke) return;
  await invoke(COMMANDS.save, { config });
}

/** メモリ上のキャッシュ（load / update 後に更新） */
let cache = null;

/**
 * 現在の設定を返す。未ロードならディスクから読む（失敗時は defaultAppConfig）。
 * @returns {Promise<AppConfig>}
 */
export async function getAppConfig() {
  if (cache) return cache;
  const loaded = await loadAppConfig();
  cache = loaded ?? defaultAppConfig();
  return cache;
}

/**
 * キャッシュを置き換える（外部で読み込んだオブジェクトを渡す場合）。
 * @param {AppConfig} config
 */
export function setAppConfigCache(config) {
  cache = config;
}

/**
 * 設定を読み、変更関数を適用し、保存する。呼び出し側はキー文字列を触らなくてよい。
 * @param {(config: AppConfig) => void} mutator
 * @returns {Promise<AppConfig>}
 */
export async function updateAppConfig(mutator) {
  const invoke = getInvoke();
  if (!invoke) {
    const c = structuredClone(await getAppConfig());
    mutator(c);
    cache = c;
    return c;
  }
  const current = await loadAppConfig();
  const base = current ?? defaultAppConfig();
  const next = structuredClone(base);
  mutator(next);
  cache = next;
  await saveAppConfig(next);
  return next;
}
